File: android-beans/src/test/java/rocks/frieler/android/beans/BeansOfTypeConsumerTest.java
package rocks.frieler.android.beans;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.junit.MockitoJUnitRunner;

import java8.util.function.Consumer;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.sameInstance;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyZeroInteractions;

@RunWith(MockitoJUnitRunner.class)
public class BeansOfTypeConsumerTest {
    @SuppressWarnings("unchecked")
    private Consumer<BeansOfTypeConsumerTest> consumer = mock(Consumer.class);
    private BeansOfTypeConsumer<BeansOfTypeConsumerTest> beansOfTypeConsumer
            = new BeansOfTypeConsumer<>(BeansOfTypeConsumerTest.class, consumer);

    @Test
    public void testPostProcessIgnoresBeanNotOfSpecifiedType() {
        final Object originalBean = new Object();

        final Object processedBean = beansOfTypeConsumer.postProcessBean("bean", originalBean);

        verifyZeroInteractions(consumer);
        assertThat(processedBean, is(sameInstance(originalBean)));
    }

    @Test
    public void testPostProcessBeanConsumesBeanOfTypeAndReturnsIt() {
        final BeansOfTypeConsumerTest processedBean = beansOfTypeConsumer.postProcessBean("bean", this);

        verify(consumer).accept(this);
        assertThat(processedBean, is(sameInstance(this)));
    }
}

File: android-beans/src/test/kotlin/rocks/frieler/android/beans/BeansOfTypeConsumerTest.kt
package rocks.frieler.android.beans

import assertk.assertThat
import assertk.assertions.isSameAs
import assertk.assertions.isSameInstanceAs
import org.junit.jupiter.api.Test
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyNoInteractions
import java.util.function.Consumer

class BeansOfTypeConsumerTest {
    private val consumer: (BeansOfTypeConsumerTest) -> Unit = mock()
    private val beansOfTypeConsumer = BeansOfTypeConsumer(BeansOfTypeConsumerTest::class, consumer)

    @Test
    fun `postProcessBean() ignores bean that is not of specified type`() {
        val originalBean = Any()

        val processedBean = beansOfTypeConsumer.postProcessBean("bean", originalBean)

        verifyNoInteractions(consumer)
        assertThat(processedBean).isSameInstanceAs(originalBean)
    }

    @Test
    fun `postProcessBean() consumes bean of specified type and returns it`() {
        val processedBean = beansOfTypeConsumer.postProcessBean("bean", this)

        verify(consumer)(this)
        assertThat(processedBean).isSameInstanceAs(this)
    }

    @Test
    fun `BeansOfTypeConsumer can be used with java-type and -Consumer`() {
        val consumer: Consumer<BeansOfTypeConsumerTest> = mock()

        val beansOfTypeConsumer = BeansOfTypeConsumer(BeansOfTypeConsumerTest::class.java, consumer)
        val processedBean = beansOfTypeConsumer.postProcessBean("bean", this)

        verify(consumer).accept(this)
        assertThat(processedBean).isSameInstanceAs(this)
    }
}
